core/job/service/job_service.go

```diff
@@ -1999,3 +1993,143 @@ func (j *JobService) getDownstreamJobs(ctx context.Context, projectName tenant.P
 	}
 	return downstreamJobs, me.ToErr()
 }
+
+func (j *JobService) BulkDeleteJobs(ctx context.Context, projectName tenant.ProjectName, jobsToDelete []*dto.JobToDeleteRequest) (map[string]dto.BulkDeleteTracker, error) {
+	// initiate deletion status per jobs to be deleted
+	deletionTrackerMap := map[string]dto.BulkDeleteTracker{}
+
+	toDeleteJobs := job.Jobs{}
+	for _, jobToDelete := range jobsToDelete {
+		existingJob, err := j.jobRepo.GetByJobName(ctx, projectName, jobToDelete.JobName)
+		if err != nil && !errors.IsErrorType(err, errors.ErrNotFound) {
+			j.logger.Error("error getting existing job %s: %s", jobToDelete.JobName.String(), err)
+			return deletionTrackerMap, err
+		}
+
+		toDeleteJobs = append(toDeleteJobs, existingJob)
+	}
+
+	toDeleteMap := toDeleteJobs.GetNameMap()
+	toDeleteSpecMap := toDeleteJobs.GetFullNameToSpecMap()
+
+	alreadyDeleted := map[job.FullName]bool{}
+	deletedJobs := job.Jobs{}
+	for _, toDeleteJob := range toDeleteJobs {
+		deletionTracker, deletedJobNames := j.resolveJobAndDownstreamsDeletion(ctx, toDeleteJob, toDeleteSpecMap, alreadyDeleted)
```

Comment:
can we also skip downstream deletion if `toDeleteJob` is already been proceeded? by checking on `deletionTrackerMap`

Reply:
see below answer